pandas/core/algorithms.py

```diff
@@ -513,60 +557,45 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
     return result
 
 
-def _value_counts_arraylike(values, dropna=True):
-    is_datetimetz_type = is_datetimetz(values)
-    is_period_type = (is_period_dtype(values) or
-                      is_period_arraylike(values))
-
-    orig = values
-
-    from pandas.core.series import Series
-    values = Series(values).values
-    dtype = values.dtype
+def _value_counts_arraylike(values, dropna):
+    """
+    Parameters
+    ----------
+    values : arraylike
+    dropna : boolean
 
-    if needs_i8_conversion(dtype) or is_period_type:
+    Returns
+    -------
+    (uniques, counts)
 
-        from pandas.tseries.index import DatetimeIndex
-        from pandas.tseries.period import PeriodIndex
+    """
+    values = _ensure_arraylike(values)
+    original = values
+    values, dtype, ndtype = _ensure_data(values)
 
-        if is_period_type:
-            # values may be an object
-            values = PeriodIndex(values)
-            freq = values.freq
+    if needs_i8_conversion(dtype):
+        # i8
 
-        values = values.view(np.int64)
         keys, counts = htable.value_count_int64(values, dropna)
 
         if dropna:
             msk = keys != iNaT
             keys, counts = keys[msk], counts[msk]
 
-        # convert the keys back to the dtype we came in
-        keys = keys.astype(dtype)
-
-        # dtype handling
-        if is_datetimetz_type:
-            keys = DatetimeIndex._simple_new(keys, tz=orig.dtype.tz)
-        elif is_period_type:
-            keys = PeriodIndex._from_ordinals(keys, freq=freq)
-
-    elif is_signed_integer_dtype(dtype):
-        values = _ensure_int64(values)
-        keys, counts = htable.value_count_int64(values, dropna)
-    elif is_unsigned_integer_dtype(dtype):
-        values = _ensure_uint64(values)
-        keys, counts = htable.value_count_uint64(values, dropna)
-    elif is_float_dtype(dtype):
-        values = _ensure_float64(values)
-        keys, counts = htable.value_count_float64(values, dropna)
     else:
-        values = _ensure_object(values)
-        keys, counts = htable.value_count_object(values, dropna)
+        # ndarray like
+
+        # TODO: handle uint8
+        f = getattr(htable, "value_count_{dtype}".format(dtype=ndtype))
```

Comment:
Not a big deal, but seems a little strange that `value_counts_...` and `duplicated_...` are free functions and not based on hashtable classes? Would unify the the dispatch.